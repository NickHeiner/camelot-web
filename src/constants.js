export const BOARD_SPACE_CLICK = 'BOARD_SPACE_CLICK';
export const MAKE_MOVE = 'MAKE_MOVE';
