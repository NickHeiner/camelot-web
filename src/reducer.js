import * as Constants from './constants';
import {fromJS} from 'immutable';
import _ from 'lodash';
import {isValidMove, applyMoves} from './utils/camelot-engine';
import {getCurrentGame, getCurrentUserPlayerName} from './utils/reducer-utils';
import {combineReducers} from 'redux';
import {firebaseStateReducer} from 'react-redux-firebase';
import {routerReducer} from 'react-router-redux';

export const defaultState = fromJS({
  chosenMoveSteps: []
});

const reducer = (state, action) => {
  switch (action.type) {
  case Constants.BOARD_SPACE_CLICK:
    return (() => {
      const chosenMoveSteps = state.getIn(['ui', 'chosenMoveSteps']);
      const boardCoords = _.pick(action.payload.boardSpace, 'row', 'col');
      const currentUserPlayerName = getCurrentUserPlayerName(state, action.payload.gameId);

      if (action.payload.boardSpace.piece && currentUserPlayerName !== action.payload.boardSpace.piece.player) {
        return state;
      }

      if (!chosenMoveSteps.size) {
        return state.setIn(['ui', 'chosenMoveSteps'], fromJS([boardCoords]));
      }

      const stepsWithThisMoveAdded = chosenMoveSteps.push(fromJS(boardCoords));
      const possibleValidMove = isValidMove(
        getCurrentGame(state, action.payload.gameId).get('gameState'),
        stepsWithThisMoveAdded, 
        currentUserPlayerName
      );

      return possibleValidMove
        ? state.setIn(['ui', 'chosenMoveSteps'], stepsWithThisMoveAdded)
        : state;
    })();
  case Constants.MAKE_MOVE:
    return (() => {
      const gameState = getCurrentGame(state, action.payload.gameId).get('gameState');
      const newGameState = applyMoves(gameState, state.getIn(['ui', 'chosenMoveSteps']));
      return state.setIn(['firebase', 'data', 'games', action.payload.gameId, 'gameState'], newGameState);
    })();
  default:
    return state;
  }
};

// Maybe this approach to combining reducers is hacky. We'll see. :)
const combinedReducer = combineReducers({
  firebase: firebaseStateReducer,
  routing: routerReducer,
  ui: state => state || null
});

export default (state, action) => {
  const nextState = combinedReducer(state, action);

  nextState.ui = nextState.ui || defaultState;

  const finalState = reducer(fromJS(nextState), action);
  return {
    firebase: finalState.get('firebase'),
    routing: finalState.get('routing').toJS(),
    ui: finalState.get('ui')
  }
};
