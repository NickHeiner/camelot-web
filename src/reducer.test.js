import reducer from './reducer';
import * as Constants from './constants';
import dataPostInitialLoad from '../fixtures/data-post-initial-load';

describe('reducer', () => {
  describe('unrecognized action', () => {
    it('is the identity function for an unrecognized action', () => {
      expect(reducer(dataPostInitialLoad, {type: 'unrecognized'})).toEqual(dataPostInitialLoad);
    });
  });

  describe(Constants.BOARD_SPACE_CLICK, () => {
    it('handles the first board space in a move chain', () => {
      const nextState = reducer(dataPostInitialLoad, {
        type: Constants.BOARD_SPACE_CLICK,
        payload: {
          gameId: '-KnrZP2chDv_4frmGLqQ',
          boardSpace: {
            col: 3,
            row: 6,
            piece: {
              player: 'playerA',
              type: 'knight'
            }
          }
        }
      });
      expect(nextState).toMatchSnapshot();
    });
  });
});
