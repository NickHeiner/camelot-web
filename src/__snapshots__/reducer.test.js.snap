exports[`reducer BOARD_SPACE_CLICK handles the first board space in a move chain 1`] = `
Object {
  "firebase": Object {
    "auth": Object {
      "displayName": "Current User Display Name",
      "photoURL": "current user photo url",
      "uid": "current-user-uid",
    },
    "authError": null,
    "data": Object {
      "games": Object {
        "-KnrZP2chDv_4frmGLqQ": Object {
          "gameState": Object {
            "boardSpaces": Array [
              Object {
                "col": 5,
                "row": 0,
              },
              Object {
                "col": 6,
                "row": 0,
              },
              Object {
                "col": 2,
                "row": 1,
              },
              Object {
                "col": 3,
                "row": 1,
              },
              Object {
                "col": 4,
                "row": 1,
              },
              Object {
                "col": 5,
                "row": 1,
              },
              Object {
                "col": 6,
                "row": 1,
              },
              Object {
                "col": 7,
                "row": 1,
              },
              Object {
                "col": 8,
                "row": 1,
              },
              Object {
                "col": 9,
                "row": 1,
              },
              Object {
                "col": 1,
                "row": 2,
              },
              Object {
                "col": 2,
                "row": 2,
              },
              Object {
                "col": 3,
                "row": 2,
              },
              Object {
                "col": 4,
                "row": 2,
              },
              Object {
                "col": 5,
                "row": 2,
              },
              Object {
                "col": 6,
                "row": 2,
              },
              Object {
                "col": 7,
                "row": 2,
              },
              Object {
                "col": 8,
                "row": 2,
              },
              Object {
                "col": 9,
                "row": 2,
              },
              Object {
                "col": 10,
                "row": 2,
              },
              Object {
                "col": 0,
                "row": 3,
              },
              Object {
                "col": 1,
                "row": 3,
              },
              Object {
                "col": 2,
                "row": 3,
              },
              Object {
                "col": 3,
                "row": 3,
              },
              Object {
                "col": 4,
                "row": 3,
              },
              Object {
                "col": 5,
                "row": 3,
              },
              Object {
                "col": 6,
                "row": 3,
              },
              Object {
                "col": 7,
                "row": 3,
              },
              Object {
                "col": 8,
                "row": 3,
              },
              Object {
                "col": 9,
                "row": 3,
              },
              Object {
                "col": 10,
                "row": 3,
              },
              Object {
                "col": 11,
                "row": 3,
              },
              Object {
                "col": 0,
                "row": 4,
              },
              Object {
                "col": 1,
                "row": 4,
              },
              Object {
                "col": 2,
                "row": 4,
              },
              Object {
                "col": 3,
                "row": 4,
              },
              Object {
                "col": 4,
                "row": 4,
              },
              Object {
                "col": 5,
                "row": 4,
              },
              Object {
                "col": 6,
                "row": 4,
              },
              Object {
                "col": 7,
                "row": 4,
              },
              Object {
                "col": 8,
                "row": 4,
              },
              Object {
                "col": 9,
                "row": 4,
              },
              Object {
                "col": 10,
                "row": 4,
              },
              Object {
                "col": 11,
                "row": 4,
              },
              Object {
                "col": 0,
                "row": 5,
              },
              Object {
                "col": 1,
                "row": 5,
              },
              Object {
                "col": 2,
                "piece": Object {
                  "player": "playerA",
                  "type": "knight",
                },
                "row": 5,
              },
              Object {
                "col": 3,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 5,
              },
              Object {
                "col": 4,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 5,
              },
              Object {
                "col": 5,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 5,
              },
              Object {
                "col": 6,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 5,
              },
              Object {
                "col": 7,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 5,
              },
              Object {
                "col": 8,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 5,
              },
              Object {
                "col": 9,
                "piece": Object {
                  "player": "playerA",
                  "type": "knight",
                },
                "row": 5,
              },
              Object {
                "col": 10,
                "row": 5,
              },
              Object {
                "col": 11,
                "row": 5,
              },
              Object {
                "col": 0,
                "row": 6,
              },
              Object {
                "col": 1,
                "row": 6,
              },
              Object {
                "col": 2,
                "row": 6,
              },
              Object {
                "col": 3,
                "piece": Object {
                  "player": "playerA",
                  "type": "knight",
                },
                "row": 6,
              },
              Object {
                "col": 4,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 6,
              },
              Object {
                "col": 5,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 6,
              },
              Object {
                "col": 6,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 6,
              },
              Object {
                "col": 7,
                "piece": Object {
                  "player": "playerA",
                  "type": "pawn",
                },
                "row": 6,
              },
              Object {
                "col": 8,
                "piece": Object {
                  "player": "playerA",
                  "type": "knight",
                },
                "row": 6,
              },
              Object {
                "col": 9,
                "row": 6,
              },
              Object {
                "col": 10,
                "row": 6,
              },
              Object {
                "col": 11,
                "row": 6,
              },
              Object {
                "col": 0,
                "row": 7,
              },
              Object {
                "col": 1,
                "row": 7,
              },
              Object {
                "col": 2,
                "row": 7,
              },
              Object {
                "col": 3,
                "row": 7,
              },
              Object {
                "col": 4,
                "row": 7,
              },
              Object {
                "col": 5,
                "row": 7,
              },
              Object {
                "col": 6,
                "row": 7,
              },
              Object {
                "col": 7,
                "row": 7,
              },
              Object {
                "col": 8,
                "row": 7,
              },
              Object {
                "col": 9,
                "row": 7,
              },
              Object {
                "col": 10,
                "row": 7,
              },
              Object {
                "col": 11,
                "row": 7,
              },
              Object {
                "col": 0,
                "row": 8,
              },
              Object {
                "col": 1,
                "row": 8,
              },
              Object {
                "col": 2,
                "row": 8,
              },
              Object {
                "col": 3,
                "row": 8,
              },
              Object {
                "col": 4,
                "row": 8,
              },
              Object {
                "col": 5,
                "row": 8,
              },
              Object {
                "col": 6,
                "row": 8,
              },
              Object {
                "col": 7,
                "row": 8,
              },
              Object {
                "col": 8,
                "row": 8,
              },
              Object {
                "col": 9,
                "row": 8,
              },
              Object {
                "col": 10,
                "row": 8,
              },
              Object {
                "col": 11,
                "row": 8,
              },
              Object {
                "col": 0,
                "row": 9,
              },
              Object {
                "col": 1,
                "row": 9,
              },
              Object {
                "col": 2,
                "row": 9,
              },
              Object {
                "col": 3,
                "piece": Object {
                  "player": "playerB",
                  "type": "knight",
                },
                "row": 9,
              },
              Object {
                "col": 4,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 9,
              },
              Object {
                "col": 5,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 9,
              },
              Object {
                "col": 6,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 9,
              },
              Object {
                "col": 7,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 9,
              },
              Object {
                "col": 8,
                "piece": Object {
                  "player": "playerB",
                  "type": "knight",
                },
                "row": 9,
              },
              Object {
                "col": 9,
                "row": 9,
              },
              Object {
                "col": 10,
                "row": 9,
              },
              Object {
                "col": 11,
                "row": 9,
              },
              Object {
                "col": 0,
                "row": 10,
              },
              Object {
                "col": 1,
                "row": 10,
              },
              Object {
                "col": 2,
                "piece": Object {
                  "player": "playerB",
                  "type": "knight",
                },
                "row": 10,
              },
              Object {
                "col": 3,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 10,
              },
              Object {
                "col": 4,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 10,
              },
              Object {
                "col": 5,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 10,
              },
              Object {
                "col": 6,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 10,
              },
              Object {
                "col": 7,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 10,
              },
              Object {
                "col": 8,
                "piece": Object {
                  "player": "playerB",
                  "type": "pawn",
                },
                "row": 10,
              },
              Object {
                "col": 9,
                "piece": Object {
                  "player": "playerB",
                  "type": "knight",
                },
                "row": 10,
              },
              Object {
                "col": 10,
                "row": 10,
              },
              Object {
                "col": 11,
                "row": 10,
              },
              Object {
                "col": 0,
                "row": 11,
              },
              Object {
                "col": 1,
                "row": 11,
              },
              Object {
                "col": 2,
                "row": 11,
              },
              Object {
                "col": 3,
                "row": 11,
              },
              Object {
                "col": 4,
                "row": 11,
              },
              Object {
                "col": 5,
                "row": 11,
              },
              Object {
                "col": 6,
                "row": 11,
              },
              Object {
                "col": 7,
                "row": 11,
              },
              Object {
                "col": 8,
                "row": 11,
              },
              Object {
                "col": 9,
                "row": 11,
              },
              Object {
                "col": 10,
                "row": 11,
              },
              Object {
                "col": 11,
                "row": 11,
              },
              Object {
                "col": 0,
                "row": 12,
              },
              Object {
                "col": 1,
                "row": 12,
              },
              Object {
                "col": 2,
                "row": 12,
              },
              Object {
                "col": 3,
                "row": 12,
              },
              Object {
                "col": 4,
                "row": 12,
              },
              Object {
                "col": 5,
                "row": 12,
              },
              Object {
                "col": 6,
                "row": 12,
              },
              Object {
                "col": 7,
                "row": 12,
              },
              Object {
                "col": 8,
                "row": 12,
              },
              Object {
                "col": 9,
                "row": 12,
              },
              Object {
                "col": 10,
                "row": 12,
              },
              Object {
                "col": 11,
                "row": 12,
              },
              Object {
                "col": 0,
                "row": 13,
              },
              Object {
                "col": 1,
                "row": 13,
              },
              Object {
                "col": 2,
                "row": 13,
              },
              Object {
                "col": 3,
                "row": 13,
              },
              Object {
                "col": 4,
                "row": 13,
              },
              Object {
                "col": 5,
                "row": 13,
              },
              Object {
                "col": 6,
                "row": 13,
              },
              Object {
                "col": 7,
                "row": 13,
              },
              Object {
                "col": 8,
                "row": 13,
              },
              Object {
                "col": 9,
                "row": 13,
              },
              Object {
                "col": 10,
                "row": 13,
              },
              Object {
                "col": 11,
                "row": 13,
              },
              Object {
                "col": 1,
                "row": 14,
              },
              Object {
                "col": 2,
                "row": 14,
              },
              Object {
                "col": 3,
                "row": 14,
              },
              Object {
                "col": 4,
                "row": 14,
              },
              Object {
                "col": 5,
                "row": 14,
              },
              Object {
                "col": 6,
                "row": 14,
              },
              Object {
                "col": 7,
                "row": 14,
              },
              Object {
                "col": 8,
                "row": 14,
              },
              Object {
                "col": 9,
                "row": 14,
              },
              Object {
                "col": 10,
                "row": 14,
              },
              Object {
                "col": 2,
                "row": 15,
              },
              Object {
                "col": 3,
                "row": 15,
              },
              Object {
                "col": 4,
                "row": 15,
              },
              Object {
                "col": 5,
                "row": 15,
              },
              Object {
                "col": 6,
                "row": 15,
              },
              Object {
                "col": 7,
                "row": 15,
              },
              Object {
                "col": 8,
                "row": 15,
              },
              Object {
                "col": 9,
                "row": 15,
              },
              Object {
                "col": 5,
                "row": 16,
              },
              Object {
                "col": 6,
                "row": 16,
              },
            ],
            "capturedPieces": Object {
              "playerA": Object {
                "knight": 0,
                "pawn": 0,
              },
              "playerB": Object {
                "knight": 0,
                "pawn": 0,
              },
            },
            "turnCount": 0,
          },
          "host": "current-user-uid",
          "opponent": "some-user-uid",
        },
      },
      "users": Object {
        "current-user-uid": Object {
          "avatarUrl": "current user photo url",
          "displayName": "Current User Display Name",
          "email": "current.user@gmail.com",
          "photoURL": "current user photo url",
          "uid": "current-user-uid",
        },
        "some-user-uid": Object {
          "avatarUrl": "other user photo url",
          "displayName": "ann perkins",
          "email": "other.user@gmail.com",
          "photoURL": "other user photo url",
          "uid": "some-user-uid",
        },
      },
    },
    "isInitializing": false,
    "profile": Object {
      "displayName": "Current User Display Name",
      "photoURL": "current user photo url",
      "uid": "current-user-uid",
    },
    "requested": Object {
      "games": true,
      "users": true,
    },
    "requesting": Object {
      "games": false,
      "users": false,
    },
    "timestamp": Object {
      "games": 1500823289327,
      "users": 1500823290264,
    },
  },
  "routing": Object {
    "locationBeforeTransitions": Object {
      "action": "POP",
      "hash": "",
      "key": "r457nc",
      "pathname": "/play/-KnrZP2chDv_4frmGLqQ",
      "query": Object {},
      "search": "",
    },
  },
  "ui": Object {
    "possibleMoveSteps": Array [
      Object {
        "col": 3,
        "row": 6,
      },
    ],
  },
}
`;
